File: src/algorithms/reverse-only-letters.js
/**
 * Given a string s, reverses the string according to the following rules:
 *   • All the characters that are not English letters remain in the same
 *     position.
 *   • All the English letters (lowercase or uppercase) should be reversed.
 * Returns s after reversing it.
 *
 * @format
 * @param {string} s - A string of characters with ASCII values in the range
 *     [33, 122]
 * @returns {string} - Returns the resulting string after reversing s
 */
function reverseOnlyLetters(s) {
  let res = "";
  const stack = [];
  for (const char of s) {
    if (charIsLetter(char)) {
      stack.push(char);
    }
  }
  for (const char of s) {
    res += charIsLetter(char) ? stack.pop() : char;
  }
  return res;
}

/**
 * A function that returns true if char is in the ASCII range [33, 122]
 *
 * @param {string} char - An ASCII character in the range [33, 122]
 * @returns {boolean} - Returns true if char is in the ASCII range [33, 122]
 */
function charIsLetter(char) {
  return /[a-zA-Z]/.test(char);
}

File: src/algorithms/reverse-only-letters.py
def reverse_only_letters(s):
    """Given a string s, reverse the string according to the following rules:
    • All the characters that are not English letters remain in the same 
    position.
    • All the English letters (lowercase or uppercase) should be reversed.
    Return s after reversing it.

    Args:
        s (str): A string of characters with ASCII values in the range [33, 122]

    Returns:
        str: Returns a modified string.
    """
    s = list(s)
    left = 0
    right = len(s) - 1
    while left < right:
        if not s[left].isalpha():
            left += 1
        elif not s[right].isalpha():
            right -= 1
        else:
            s[right], s[left] = s[left], s[right]
            left += 1
            right -= 1
    return "".join(s)
